### PR TITLE
SALTO-4599 - Zendesk Adapter - Add custom_object

### PR DESCRIPTION
**Pull request by idoser96**
Add support for fetching custom_objects in Zendesk
* Creating the new types
* Reference expressions
* UT E2E

---

---
_Release Notes_: 
Zendesk Adapter:
* Added support to fetch custom_object instances

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_field_options.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  ZENDESK,
+} from '../../constants'
+
+const { RECORDS_PATH } = elementsUtils
+const log = logger(module)
+
+export const customObjectFieldOptionType = new ObjectType({
+  elemID: new ElemID(ZENDESK, CUSTOM_OBJECT_FIELD_OPTIONS_TYPE_NAME),
+  fields: {
+    id: { refType: BuiltinTypes.SERVICE_ID_NUMBER, annotations: { [CORE_ANNOTATIONS.HIDDEN_VALUE]: true } },
```
Comment (idoser96) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_field_options.ts:42]: why doesn't this work?

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+const { makeArray } = collections.array
+const { createMissingInstance } = referencesUtils
+const { createPaginator } = clientUtils
+
+const relationTypeToType = (relationshipTargetType: Value): string => {
+  if (!_.isString(relationshipTargetType)) {
+    return 'unknown'
+  }
+
+  if (relationshipTargetType.startsWith('zen:custom_object')) {
```
Comment (idoser96) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:54]: // This scenario seems possible, but I did not manage to create it in the UI

File: packages/zendesk-adapter/src/filters/custom_objects/utils.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    (firstInstance: string | ReferenceExpression, secondInstance: string | ReferenceExpression) => TemplateExpression
+}): TemplateExpression | string => {
+  if (!enableMissingReferences) {
+    return buildTemplateFunc(ticketField, customKey).parts.join()
+  }
+  const missingInstance = createMissingInstance(ZENDESK, missingInstanceType, missingInstanceName)
+  const missingInstanceRef = new ReferenceExpression(missingInstance.elemID)
+  return isReferenceExpression(ticketField)
+    ? buildTemplateFunc(ticketField, missingInstanceRef)
+    : buildTemplateFunc(missingInstanceRef, customKey)
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/utils.ts:58]: why here you don't do parts.join()?

File: packages/zendesk-adapter/src/filters/custom_objects/utils.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+  : TemplateExpression => new TemplateExpression({
+  parts: [
+    'lookup:ticket.ticket_field_',
+    ticketField,
+    '.custom_fields.',
+    option,
+  ],
+})
+
+export const transformCustomObjectLookupField = ({
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/custom_objects/utils.ts:77]: can you add a specific test for this?

**Review by edenhassid — COMMENTED** [review]
good work!! 👏 🤩 we can go over the comment and the code IRL

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+  usersById,
+}: {
+  customObjectField?: InstanceElement
+  filter: { operator: string; value?: string | number | ReferenceExpression }
+  enableMissingReferences: boolean
+  instancesById: Record<string, InstanceElement>
+  usersById: Record<string, string>
+}): void => {
+  if (customObjectField === undefined
+    // These are special cases where the value is a reference to an element
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:84]: The special case is when it is a reference

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+type CustomObjectAction = {
+  field: string
+  value: Array<string | TemplateExpression> | string | TemplateExpression
+}
+
+const isRelevantAction = (action: Value): action is CustomObjectAction =>
+  _.isPlainObject(action)
+  && _.isString(action.field)
+  // makeArray to catch both cases where value is a string or an array of strings
+  && makeArray(action.value).every(_.isString)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:135]: Added a comment on the type

File: packages/zendesk-adapter/src/filters/custom_objects/utils.ts (hunk 4 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/utils.ts:77]: Talked IRL, added documentation

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts (hunk 2 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:57]: cant put 'startsWith' in switch

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  actions
+    .filter(isRelevantAction)
+    .forEach(action => {
+      const value = makeArray(action.value)[0]
+      // always false, used for type casting
+      if (!_.isString(value)) {
+        return
+      }
+      const { result } = transformCustomObjectLookupField({ field: value, ...transformArgs })
+      // notification_user is a special case, value is an array and the first element is the custom_object field
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts]: There might be but i did not manage to create it in the UI, and i have no idea what it might be if there is

**Review by edenhassid — COMMENTED** [review]
cool! 😎

**Review by edenhassid — DISMISSED** [review]
LGTM! 🔥 🚀  just make sure to figure out if we need a noise reduction or not before you merge.

File: packages/adapter-utils/src/template.ts [hunk 8]
```diff
@@ -37,10 +37,17 @@ export const compactTemplateParts = (parts: TemplatePart[]): TemplatePart[] => {
       compactedParts.push(part)
     }
   })
-  compactedParts.push(tempString.join(''))
+  if (tempString.length > 0) {
```
Comment (netama) [inline packages/adapter-utils/src/template.ts:40]: is this related?

File: packages/adapter-utils/src/template.ts [hunk 9]
```diff
@@ -37,10 +37,17 @@ export const compactTemplateParts = (parts: TemplatePart[]): TemplatePart[] => {
...
+    compactedParts.push(tempString.join(''))
+  }
   return compactedParts
 }
 
+export const compactTemplate = (template: TemplateExpression): TemplateExpression | string => {
+  const parts = compactTemplateParts(template.parts)
+  return parts.every(_.isString) ? parts.join() : new TemplateExpression({ parts })
```
Comment (netama) [inline packages/adapter-utils/src/template.ts:48]: do we have to add this one? I'm not sure it belongs as a utility function, we're already using the existing function in `createTemplateExpression` which was the agreed pattern. can't we just create the template expression with `createTemplateExpression` in the new usage as well?

File: packages/zendesk-adapter/src/config.ts [hunk 10]
```diff
@@ -2488,6 +2488,124 @@ export const DEFAULT_TYPES: ZendeskApiConfig['types'] = {
       isSingleton: true,
     },
   },
+  custom_objects: {
+    request: {
+      url: '/api/v2/custom_objects',
+      queryParams: { per_page: String(PAGE_SIZE) },
```
[the remark this answers is not in the document]
Comment (netama) [reply packages/zendesk-adapter/src/config.ts:2494]: let's add some TODO to switch when available (and maybe ask when that will be - @edenhassid probably best that you ask)?

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_field_options.ts [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+  _.isArray(options) && options.every(option =>
+    _.isPlainObject(option)
+    && _.isNumber(option.id)
+    && _.isString(option.name)
+    && _.isString(option.raw_name)
+    && _.isString(option.value))
+
+/**
+ * Convert custom_field_options of custom_object_field to be instance elements
+ * This is needed because 'extractStandaloneFields' doesn't support types from 'recurse into'
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_field_options.ts:70]: wdym? we do this in a few places, what's different here? (can discuss IRL)

File: packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+  field: string | TemplateExpression
+  operator: string
+  value?: string | ReferenceExpression
+}
+
+const isRelevantRule = (rule: Value): rule is Rule =>
+  _.isPlainObject(rule)
+  && _.isString(rule.operator)
+  && (rule.value === undefined || _.isString(rule.value))
+  && RELATIONSHIP_FILTER_REGEX.test(rule.field)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/custom_objects/custom_object_fields.ts:83]: please surround with parens when there's a multiline implementation
```suggestion
const isRelevantRule = (rule: Value): rule is Rule => (
  _.isPlainObject(rule)
  && _.isString(rule.operator)
  && (rule.value === undefined || _.isString(rule.value))
  && RELATIONSHIP_FILTER_REGEX.test(rule.field)
)
```

File: packages/adapter-utils/src/template.ts (hunk 8 above)
Comment (idoser96) [reply packages/adapter-utils/src/template.ts:40]: a fix of a bug that i found during my testings
without that if, if a template expression is ending with a reference expression, it will have an empty string in the last part

File: packages/adapter-utils/src/template.ts (hunk 9 above)
Comment (idoser96) [reply packages/adapter-utils/src/template.ts:48]: I created this function to handle a situation when all the parts are strings, so we need to return a string
changing 'createTemplateExpression' will require checking all its usages
using createTemplateExpression will require checking if its only strings after each usage

so I created this function to do it automatically, i can change it if you still think it is better

**Review by edenhassid — APPROVED** [review]
great!! 🥳 💯 🔥
